src/tag_gc.rs

```diff
@@ -3,30 +3,120 @@ use rustc_data_structures::fx::FxHashSet;
 use crate::*;
 
 pub trait VisitMachineValues {
-    fn visit_machine_values(&self, visit: &mut impl FnMut(&Operand<Provenance>));
+    fn visit_machine_values(&self, visit: &mut ProvenanceVisitor);
+}
```

Comment:
Hm okay this is less general than what I started out with, but I guess for now visiting all provenance is sufficient.

But then it should probably be called `VistiProvenance`.
And why did you change the `visit` argument?

Reply:
We need to visit a number of different things. It's possible to contort them all into an `Operand`, but that adds a lot of repetitive glue code at the call sites. If each use of the visitor can take a different type, we can do the dispatch inside this module, instead of spread across the codebase. So we need to shift the generic parameter to each call to the visitor, which means we need a new type with the generic parameter on the method, not `FnMut`.

Reply:
But it seems you solved this by completely re-structuring the visitor to (a) to longer be lazy (it now hard-codes the use of a hash map -- `ProvenanceVisitor` is a somewhat misleading name) and (b) hard-code that only provenance matters.

If we want to hard-code that only provenance matters (which I am fine with for now), we should have
```rust
pub trait VisitProvenance {
 fn visit_provenance(&self, visit: &mut impl FnMut(SbTag));
}
```
Then you can `impl VisitProvenance for {Operand,Immediate,Pointer}` and you can now do `x.visit_provenance(visit)` for `x ∈ {Operand,Immediate,Pointer}`.

If we don't want to hard-code hat only provenance matters we need something else, but your approach does not achieve that.

Reply:
That said, some places only have `SbTag`, not a full `Provenance`... not sure what the best way is to handle them. We could make the visitor focus only on `SbTag` I guess; then it should be called `VisitProvenanceTags` or so.

Reply:
I pushed to this branch a version of these traits (or rather, just a single trait) demonstrating what I meant.